black/white e-Paper/e-Ink display driver

silkeh opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR adds a driver for black/white e-paper/e-ink SPI display drivers/controllers, including:

- Default parameters for the IL3829 (tested)
- Default parameters for the SSD1673 (see #11096, untested)
- Header files containing RIOT logos. These were generated using [png2c][2].
  Some questions:
   - Should I move these to a generally available directory?
   - What should the copyright notice be?
- Tests for the display driver. A video with the full test can be found [here][3].

With the right parameters, this driver should work with most black and white e-paper display drivers. This has only been tested with the IL3829, which is commonly used in 1.54 inch black and white e-Paper displays like [this one][1].

This PR is based on #11078 with some major rework to get it all in a more generic package.

### Testing procedure
A test program is included (`tests/driver_il3829`).
Modify the makefile for a microcontroller/display combination you have and go for it.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
This is a complete rework based on the feedback from #11078.

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

[1]: https://www.waveshare.com/product/modules/oleds-lcds/e-paper/1.54inch-e-paper-module.htm
[2]: https://github.com/silkeh/png2c
[3]: https://slxh.nl/files/il3829_demo.webm
[4]: https://github.com/RIOT-OS/RIOT/pull/9948/files#r225003782

silkeh commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,290 @@
+    return 0;
+}
+
+void epd_bw_spi_display_init(epd_bw_spi_t *dev)
```
@Citrullin made a good point in the original implementation:
> The factory firmware of good-display pulls the RST pin down and up here. I would use this here too. The wake up function would make sense here then.

The reason I did not do this is that the display does not have to be in sleep here. One display I have does not actually have a reset pin, for example. I see the case for calling `epd_bw_spi_wake` here though, as it doesn't hurt an awake display.

silkeh commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,290 @@
+{
+    DEBUG("[epd_bw_spi] cmd_start: command 0x%02x\n", cmd);
+    if (p->busy_pin != GPIO_UNDEF) {
+        while (gpio_read(p->busy_pin)) {}
```
@Citrullin noted:

> We should change this. The IL0373 is active low on BUSY, while the IL3829 is active high.

I could add an expected value to the parameters so we can compare this.

Citrullin replied:
Sounds good.

silkeh commented:
@bergzand ping!

bergzand commented:
@aabadie You have a proposal on a unified display driver API right? Would it be possible for you to give this driver a review at some point?

Citrullin commented:
@bergzand Where is the proposal for the driver. I would like to take a look into it, if it suits my needs as well :)

bergzand commented:
@Citrullin There is a start of it in #13124 but I don't think there is a dedicated PR or issue yet

aabadie commented:
> Where is the proposal for the driver?

It's in e3fb86eb05ff45cf4fae88f3e1096bf800ce3587 and e7cbd70ce1de4666c3fb184670949f97bebb126a from #13124 but it's still very minimal (only one map function exposed by the generic API). I'm planning to add getter/setter functions to be able to control/access internal states/parameters of the display drivers from higher level functions (like lvgl for example).
If you think the API is useful, I can split it out from #13124 and open a dedicated PR.

silkeh commented:
I have updated the PR with an implementation of `disp_dev`, some minor fixes and a rebase on master.

Citrullin commented:
@silkeh I will start working on it as well. I need to split up your code in order to make it work with the IL0373. For example the variable values in `epd_bw_spi_internal.h` are different for the IL0373. There are some fixed values, which needs to get variables.

silkeh commented:
@aabadie, @bergzand: ping! :birthday:

jeandudey commented on drivers/epd_bw_spi/include/epd_bw_spi_params.h:
```diff
@@ -0,0 +1,129 @@
+#define EPD_BW_SPI_PARAM_SPI_CLK    (SPI_CLK_5MHZ)
+#endif
+#ifndef EPD_BW_SPI_PARAM_CS
+#define EPD_BW_SPI_PARAM_CS         (GPIO_UNDEF)
```
```suggestion
#define EPD_BW_SPI_PARAM_CS         (SPI_CS_UNDEF)
```

I think `SPI_CS_UNDEF` could be used here

jeandudey commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,321 @@
+        while (gpio_read(p->busy_pin)) {}
+    }
+    else {
+        DEBUG("[epd_bw_spi] wait: for %lu microseconds\n", usec);
```
```suggestion
        DEBUG("[epd_bw_spi] wait: for %"PRIu32" microseconds\n", usec);
```

jeandudey commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,321 @@
+void epd_bw_spi_init_full(epd_bw_spi_t *dev)
+{
+    epd_bw_spi_display_init(dev);
+    epd_bw_spi_set_area(dev, 0, 200, 0, 200);
```
Shouldn't it be `dev->size_`? :eyes:

Also should this be called before setting the look-up table? In various drivers for Waveshare e-Paper displays they seem to do it later. The one I'm using to test is a [1.54 in e-Ink](https://www.waveshare.com/wiki/1.54inch_e-Paper_Module) with source code [here](https://github.com/waveshare/e-Paper/) (I haven't made it work with this driver, yet).

silkeh replied:
Yes, it should be.

I don't think the initialization order here matters. The most important thing is that both registers are initialized. I have tested this driver with two displays, one of them being the Waveshare display and the other one being one with a missing busy/reset pin.

jeandudey commented on tests/driver_epd_bw_spi_disp_dev/Makefile:
```diff
@@ -0,0 +1,17 @@
+include ../Makefile.tests_common
+
+BOARD := nucleo-f411re
```
```suggestion
BOARD ?= nucleo-f411re
```

To allow using other boards :wink:

jeandudey commented on tests/driver_epd_bw_spi/Makefile:
```diff
@@ -0,0 +1,16 @@
+include ../Makefile.tests_common
+
+BOARD := nucleo-f411re
```
dito

jeandudey commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,321 @@
+
+void epd_bw_spi_wake(epd_bw_spi_t *dev)
+{
+    /* Give a low pulse on the reset pin */
+    if (dev->params.rst_pin != GPIO_UNDEF) {
+        gpio_clear(dev->params.rst_pin);
+        xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
+        gpio_set(dev->params.rst_pin);
+    }
```
What about making this a function akin to `epd_bw_spi_swreset` like:

```c
void epd_bw_spi_hwreset(epd_bw_spi_t *dev)
{
    /* Give a low pulse on the reset pin */
    if (dev->params.rst_pin != GPIO_UNDEF) {
        gpio_clear(dev->params.rst_pin);
        xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
        gpio_set(dev->params.rst_pin);
        xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
    }
}
```

With an additional 1 ms wait time used/needed for SSD1608 IC drivers, this way to reset the device at `epd_bw_spi_init_[full/part]` (as done for SSD1608) and should do no harm for other ICs that don't have a reset pin nor are required to be reset at initialization.

jeandudey replied:
_Btw digging up in the deepest of the internet I've found a [datasheet for SSD1608](https://annejan.com/media/SSD1608_1.2.pdf) used on Waveshare 1.54 in e-Paper devices_

silkeh replied:
Thanks, changed!

Also: thanks for the datasheet!

jeandudey commented:
Finally made it work on an `esp32-wroom-32` board with the following patch (with a previous `git rebase master`):

<details>
<summary>0001-fixup-epd_bw_spi-add-generic-black-and-white-SPI-e-P.patch</summary>

```patch
From d7a2b5f36b52b13970758371017d6d6f559d451c Mon Sep 17 00:00:00 2001
From: Jean Pierre Dudey <me@jeandudey.tech>
Date: Mon, 2 Nov 2020 11:14:47 +0100
Subject: [PATCH] fixup! epd_bw_spi: add generic black and white SPI e-Paper
 display driver

---
 drivers/epd_bw_spi/epd_bw_spi.c               | 23 +++++++++++-------
 .../epd_bw_spi/include/epd_bw_spi_params.h    | 24 ++++++++++++++++++-
 drivers/include/epd_bw_spi.h                  |  7 ++++++
 3 files changed, 45 insertions(+), 9 deletions(-)

diff --git a/drivers/epd_bw_spi/epd_bw_spi.c b/drivers/epd_bw_spi/epd_bw_spi.c
index c1c407a40..6f76537a3 100644
--- a/drivers/epd_bw_spi/epd_bw_spi.c
+++ b/drivers/epd_bw_spi/epd_bw_spi.c
@@ -64,7 +64,7 @@ void epd_bw_spi_wait(epd_bw_spi_params_t *p, uint32_t usec)
         while (gpio_read(p->busy_pin)) {}
     }
     else {
-        DEBUG("[epd_bw_spi] wait: for %lu microseconds\n", usec);
+        DEBUG("[epd_bw_spi] wait: for %"PRIu32"microseconds\n", usec);
         xtimer_usleep(usec);
     }
 }
@@ -114,9 +114,10 @@ int epd_bw_spi_init(epd_bw_spi_t *dev,
 
 void epd_bw_spi_display_init(epd_bw_spi_t *dev)
 {
+    epd_bw_spi_hwreset(dev);
+
     le_uint16_t y_data[2] = { 0 };
     uint8_t y_size;
-
     if (dev->controller.size_y <= 255) {
         y_data[0].u8[0] = dev->size_y - 1;
         y_size = 2;
@@ -179,7 +180,7 @@ void epd_bw_spi_deactivate(epd_bw_spi_t *dev)
 void epd_bw_spi_init_full(epd_bw_spi_t *dev)
 {
     epd_bw_spi_display_init(dev);
-    epd_bw_spi_set_area(dev, 0, 200, 0, 200);
+    epd_bw_spi_set_area(dev, 0, dev->size_x, 0, dev->size_y);
     epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_WRITE_LUT_REGISTER,
                          dev->controller.lut_full, dev->controller.lut_size);
 }
@@ -301,18 +302,24 @@ void epd_bw_spi_sleep(epd_bw_spi_t *dev)
 }
 
 void epd_bw_spi_wake(epd_bw_spi_t *dev)
+{
+    epd_bw_spi_hwreset(dev);
+
+    /* Turn off sleep mode */
+    uint8_t data[] = { 0x00 };
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DEEP_SLEEP_MODE, data, 1);
+    epd_bw_spi_wait(&dev->params, EPD_BW_SPI_WAIT_RESET);
+}
+
+void epd_bw_spi_hwreset(epd_bw_spi_t *dev)
 {
     /* Give a low pulse on the reset pin */
     if (dev->params.rst_pin != GPIO_UNDEF) {
         gpio_clear(dev->params.rst_pin);
         xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
         gpio_set(dev->params.rst_pin);
+        xtimer_usleep(EPD_BW_SPI_WAIT_RESET);
     }
-
-    /* Turn off sleep mode */
-    uint8_t data[] = { 0x00 };
-    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DEEP_SLEEP_MODE, data, 1);
-    epd_bw_spi_wait(&dev->params, EPD_BW_SPI_WAIT_RESET);
 }
 
 void epd_bw_spi_swreset(epd_bw_spi_t *dev)
diff --git a/drivers/epd_bw_spi/include/epd_bw_spi_params.h b/drivers/epd_bw_spi/include/epd_bw_spi_params.h
index a0590ff40..544ad0a27 100644
--- a/drivers/epd_bw_spi/include/epd_bw_spi_params.h
+++ b/drivers/epd_bw_spi/include/epd_bw_spi_params.h
@@ -37,7 +37,7 @@ extern "C" {
 #define EPD_BW_SPI_PARAM_SPI_CLK    (SPI_CLK_5MHZ)
 #endif
 #ifndef EPD_BW_SPI_PARAM_CS
-#define EPD_BW_SPI_PARAM_CS         (GPIO_UNDEF)
+#define EPD_BW_SPI_PARAM_CS         (SPI_CS_UNDEF)
 #endif
 #ifndef EPD_BW_SPI_PARAM_DC
 #define EPD_BW_SPI_PARAM_DC         (GPIO_UNDEF)
@@ -84,6 +84,15 @@ static const uint8_t epd_bw_spi_il3829_lut_default_part[] = {
     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
     0x13, 0x14, 0x44, 0x12, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
 
+/**
+ * @brief   Waveform lookup table for a full display refresh for SSD1608
+ *          (HINK-E0154A05 e-Paper).
+ */
+static const uint8_t epd_bw_spi_ssd1608_hink_e0154a05_lut_default_full[] = {
+    0x02, 0x02, 0x01, 0x11, 0x12, 0x12, 0x22, 0x22, 0x66, 0x69,
+    0x69, 0x59, 0x58, 0x99, 0x99, 0x88, 0x00, 0x00, 0x00, 0x00,
+    0xF8, 0xB4, 0x13, 0x51, 0x35, 0x51, 0x51, 0x19, 0x01, 0x00 };
+
 /**
  * @brief    Configuration for IL3829 e-paper display controller
  */
@@ -108,6 +117,19 @@ static const uint8_t epd_bw_spi_il3829_lut_default_part[] = {
         .lut_part = epd_bw_spi_il3829_lut_default_part, \
 }
 
+/**
+ * @brief   Configuration for SSD1608 (HINK-E0154A05) e-paper display controller
+ */
+#define EPD_BW_SPI_CONTROLLER_SSD1608_HINK_E0154A05 { \
+        .vcom = 0xA8, \
+        .size_x = 240, \
+        .size_y = 320, \
+        .lut_size = sizeof epd_bw_spi_ssd1608_hink_e0154a05_lut_default_full, \
+        .lut_full = epd_bw_spi_ssd1608_hink_e0154a05_lut_default_full, \
+        /* NOTE: uses this one of IL3829 for partial updates */ \
+        .lut_part = epd_bw_spi_il3829_lut_default_part, \
+}
+
 /**
  * @brief    Default controller definition
  */
diff --git a/drivers/include/epd_bw_spi.h b/drivers/include/epd_bw_spi.h
index a64165771..8f6767ed3 100644
--- a/drivers/include/epd_bw_spi.h
+++ b/drivers/include/epd_bw_spi.h
@@ -233,6 +233,13 @@ void epd_bw_spi_sleep(epd_bw_spi_t *dev);
  */
 void epd_bw_spi_wake(epd_bw_spi_t *dev);
 
+/**
+ * @brief   Perform a hardware reset of the device.
+ *
+ * This will make the IC registers to be on POR values.
+ */
+void epd_bw_spi_hwreset(epd_bw_spi_t *dev);
+
 /**
  * @brief   Perform a soft reset of the device.
  *
-- 
2.17.1

```
</details>

With the example code of `tests/driver_epd_bw_spi`:


```make
include ../Makefile.tests_common

BOARD ?= nucleo-f411re

USEMODULE += xtimer
USEMODULE += epd_bw_spi

CFLAGS += -DDISPLAY_X=200
CFLAGS += -DDISPLAY_Y=200
CFLAGS += -DEPD_BW_SPI_CONTROLLER=EPD_BW_SPI_CONTROLLER_SSD1608_HINK_E0154A05
CFLAGS += -DEPD_BW_SPI_PARAM_CS=GPIO23
CFLAGS += -DEPD_BW_SPI_PARAM_DC=GPIO16
CFLAGS += -DEPD_BW_SPI_PARAM_RST=GPIO17
CFLAGS += -DEPD_BW_SPI_PARAM_BUSY=GPIO21

CFLAGS += -DSPI0_MOSI=GPIO18
CFLAGS += -DSPI0_SCK=GPIO5
CFLAGS += -DSPI0_CS=GPIO23

include $(RIOTBASE)/Makefile.include
```

![image](https://user-images.githubusercontent.com/5952531/97857002-631bc500-1cfd-11eb-9386-d229fbef3bc7.png)

Citrullin commented:
@jeandudey This is still an open topic, because it makes more sense to have an unified display API for this. It would be great to get some input on it: #14054 and #14051

aabadie commented on tests/driver_epd_bw_spi/Makefile:
```diff
@@ -0,0 +1,16 @@
+
+CFLAGS += -DDISPLAY_X=200
+CFLAGS += -DDISPLAY_Y=200
+CFLAGS += -DEPD_BW_SPI_CONTROLLER=EPD_BW_SPI_CONTROLLER_IL3829
+CFLAGS += -DEPD_BW_SPI_PARAM_CS=GPIO_PIN\(PORT_B,6\)
+CFLAGS += -DEPD_BW_SPI_PARAM_DC=GPIO_PIN\(PORT_C,7\)
+CFLAGS += -DEPD_BW_SPI_PARAM_RST=GPIO_PIN\(PORT_A,8\)
+CFLAGS += -DEPD_BW_SPI_PARAM_BUSY=GPIO_PIN\(PORT_B,10\)
```
I would not hardcode all these parameters in a generic application Makefile. I have an STM32 board with a compatible screen (and a branch ;) ) where I would like to define all these params in the board configuration. The problem with the approach here is that it will conflict.

silkeh replied:
Do you have a suggestion about where to define it for people (like me) that don't have a board with a display?

jeandudey replied:
On projects that require custom config for a board I usually do create a `board.h` fine and `#include_next "board.h"` on it. E.g:

- board.h
```c
#ifdef BOARD_BOARD_NUCLEO_F411RE
#define EPD_BW_SPI_CONTROLLER=EPD_BW_SPI_CONTROLLER_IL3829
#define EPD_BW_SPI_PARAM_CS=GPIO_PIN(PORT_B,6)
#define EPD_BW_SPI_PARAM_DC=GPIO_PIN(PORT_C,7)
#define EPD_BW_SPI_PARAM_RST=GPIO_PIN(PORT_A,8)
#define EPD_BW_SPI_PARAM_BUSY=GPIO_PIN(PORT_B,10)
#endif

#include_next "board.h"
```

And add to applications Makefile `INCLUDES += -I$(APPDIR)`.

jeandudey replied:
See also [this](https://doc.riot-os.org/group__cpu__esp8266.html#esp8266_application_specific_board_configuration) and [this](https://doc.riot-os.org/group__cpu__esp8266.html#esp8266_application_specific_driver_configuration) which applies to all boards I think.

aabadie replied:
The usual way in RIOT for this kind of driver parameters is to define them (with default values) in `<driver>_params.h`. Then, if a board requires specific parameters, they can be defined in the board.h header of the board.

See the [driver implementation guidelines](http://doc.riot-os.org/driver-guide.html), see the `Default device configuration` paragraph where this is explained.

aabadie commented on tests/driver_epd_bw_spi/Makefile:
```diff
@@ -0,0 +1,16 @@
+USEMODULE += xtimer
+USEMODULE += epd_bw_spi
+
+CFLAGS += -DDISPLAY_X=200
+CFLAGS += -DDISPLAY_Y=200
```
The DISPLAY_X and DISPLAY_Y values should be provided as params of the driver.

silkeh replied:
Do you mean I should add a separate list of defines for the x/y size of various displays?

silkeh commented:
> @silkeh I started working on it as well. I need to split up your code in order to make it work with the IL0373. For example the variable values in `epd_bw_spi_internal.h` are different for the IL0373.

The IL0373 is not black and white (but tricolor), so it would be out of scope for this driver. Let's try to fix that in a future PR, shall we?

> Finally made it work on an esp32-wroom-32 board with the following patch

Thanks for taking the time to test and review this. I have added a define for the 1608, are you sure the different LUT is required? The one for the IL3829 seems to work fine with my WaveShare 1.54 inch display.

I think these questions may have been missed in my PR description:

>  - Header files containing RIOT logos. These were generated using png2c.
>   Some questions:
>    - Should I move these to a generally available directory?
>    - What should the copyright notice be?

Bonus picture:

![DSC_1360](https://user-images.githubusercontent.com/5798032/108417285-0ef70b00-7230-11eb-99b9-b105b764e568.JPG)

jeandudey commented:
>Thanks for taking the time to test and review this. I have added a define for the 1608, are you sure the different LUT is required? The one for the IL3829 seems to work fine with my WaveShare 1.54 inch display.

Seems to be the recommended by Waveshare on their Arduino driver, not an strong opinion though. IIRC I had errors without the Waveshare LUT, e.g. something along the lines of noise on the screen on a full update.

jeandudey commented on tests/driver_epd_bw_spi_disp_dev/main.c:
```diff
@@ -0,0 +1,75 @@
+        xtimer_periodic_wakeup(&last_wakeup, 5 * US_PER_SEC);
+
+        /* Use partial refreshes for the small logo */
+        /* cppcheck-suppress redundantAssignment
+         * (reason: this variable is checked in _epd_bw_spi_map by casting
+         *  the disp_dev_t back to epd_bw_spi_t) */
+        epd.refresh_partial = true;
```
Maybe adding an inline function would be helpful for others if running on the same problem, maybe something similar to `epd_bw_spi_set_refresh_partial`.

silkeh replied:
I have added `epd_bw_spi_set_partial`

Citrullin commented:
> > @silkeh I started working on it as well. I need to split up your code in order to make it work with the IL0373. For example the variable values in `epd_bw_spi_internal.h` are different for the IL0373.
> 
> The IL0373 is not black and white (but tricolor), so it would be out of scope for this driver. Let's try to fix that in a future PR, shall we?
> 

If it isn't a generic driver for all, or at least most, epaper displays, I feel quite uncomfortable to merge this into mainline at the current state. It would probably better to change the whole naming convention to IL3829. We also still have an open discussion about abstracting the interface for displays in general. That probably will also influence they way we write drivers. I would like to see a similar abstraction as we have with SAUL.

aabadie commented:
> I would like to see a similar abstraction as we have with SAUL.

[disp_dev](https://doc.riot-os.org/group__drivers__disp__dev.html) is meant for that and should work with PR. Someone should review/discuss #14054, #14051 and #13787. Regarding #14051, IIRC there was an issue with code size increase but needs to be investigated.

jeandudey commented:
>Should I move these to a generally available directory?

I think they should be on their example/test application since providing it in a central place would probably need to account for different sizes and scales, at the end of the day it's easier to manually generate an image using png2c like you did.

>What should the copyright notice be?

Probably the original license of the logo as it's machine generated code of the official RIOT logo, it has the same intellectual property of their owners/creators. I don't know what license it does have though :( .

fjmolinas commented:
ping @silkeh :)

silkeh commented:
@fjmolinas Will rework if somebody can tell me what the license is on the RIOT logo (or who holds the copyright).

waehlisch commented:
we clarified the usage of the RIOT logo on the new website (temp version here https://relaunch.riot-os.org/branding.html). As copyright notice you link to https://riot-os.org/branding.html (will be available by the end of this week.

silkeh commented:
Updated with:

- Logos next to the examples
- `epd_bw_spi_set_partial` for switching the display between partial and full refresh mode
- Board override in test files in `epd_bw_spi_params.h`

As for the feedback about X/Y-coordinates: if it is preferable to set the display parameters (instead of only the display driver parameters) in the driver it would probably require some modification of the parameters into with something like this:

```c
/**
 * @brief   Display parameters.
 */
typedef struct {
    uint16_t size_x;                    /**< number of horizontal pixels in the display */
    uint16_t size_y;                    /**< number of vertical pixels in the display */
    epd_bw_spi_entry_mode_t entry_mode; /**< data entry mode */
} epd_bw_spi_display_t;

/**
 * @brief   Driver parameters.
 */
typedef struct {
    epd_bw_spi_interface_t spi;         /**< SPI display parameters */
    epd_bw_spi_controller_t controller; /**< E-paper controller configuration */
    epd_bw_spi_display_t display;       /**< Display parameters */
} epd_bw_spi_params_t;
```

Which I can do in the next week or so, if it's desired.

aabadie commented on drivers/include/epd_bw_spi.h:
```diff
@@ -0,0 +1,251 @@
+ * @brief   Device initialisation parameters.
+ */
+typedef struct {
+    #ifdef MODULE_DISP_DEV
+    disp_dev_t *dev;                    /**< Pointer to the generic display device */
+    bool refresh_partial;               /**< Use partial instead of full refreshes */
+    #endif
```
```suggestion
#ifdef MODULE_DISP_DEV
    disp_dev_t *dev;                    /**< Pointer to the generic display device */
#endif
```

See my comment above about the partial_refresh attribute.

aabadie commented on drivers/include/epd_bw_spi_disp_dev.h:
```diff
@@ -0,0 +1,45 @@
+ * @brief   Reference to the display device driver struct
+ */
+extern const disp_dev_driver_t epd_bw_spi_disp_dev_driver;
+
+/**
+* @brief   Configure the display to update with a full or partial refresh
+*
+* @param[in] dev     Device descriptor.
+* @param[in] partial Boolean to set partial or full mode.
+*/
+void epd_bw_spi_set_partial(disp_dev_t *dev, bool partial);
```
```suggestion
```

Same comment about `partial_refresh`.

aabadie commented on tests/driver_epd_bw_spi/Makefile:
```diff
@@ -0,0 +1,10 @@
+USEMODULE += xtimer
+USEMODULE += epd_bw_spi
+
+INCLUDES += -I$(APPDIR)
```
This is a configuration for your personal setup. I don't think it should be put there (I mean in RIOT). Maybe someone with a nucleo-f411re won't wire the same e-ink paper display than yours and maybe not on the same pins.
So I would simply remove the `epd_bw_spi_params.h` include from the application directory (it should remain generic as much as possible). Note that this would simply forbid the possibility to build this application with a board providing a e-ink paper: the board support won't be able to provide custom defines (different pins, screen size, controller) the usual way.
The x/y display sizes and controller are the same as the default ones, so no need to overwrite them here.

Also note that I would be ok to move the setup specific pin definitions to the default driver params with a comment (use the same one used in the application params.h file, you can even generalize to any Nucleo 64 board I guess, they have a compatible pinout).

silkeh replied:
> This is a configuration for your personal setup. I don't think it should be put there (I mean in RIOT). 

Sure, but it's the only setup I have to test with.

> Maybe someone with a nucleo-f411re won't wire the same e-ink paper display than yours and maybe not on the same pins.

Which is why the pinout is documented...

> So I would simply remove the epd_bw_spi_params.h include from the application directory (it should remain generic as much as possible).

Sure, but that just means that I have to keep it untracked in my code base. Otherwise I won't be able to actually run this application.

> Note that this would simply forbid the possibility to build this application with a board providing a e-ink paper: the board support won't be able to provide custom defines (different pins, screen size, controller) the usual way.

How would removing the header file lead to this application not working with a board having an e-paper display?

> The x/y display sizes and controller are the same as the default ones, so no need to overwrite them here.

I'd much rather have a complete override instead of a partial one in case defaults change.

> Also note that I would be ok to move the setup specific pin definitions to the default driver params with a comment (use the same one used in the application params.h file, you can even generalize to any Nucleo 64 board I guess, they have a compatible pinout).

Sorry, but I don't think it makes any sense: my personal setup can't be included explicitly, but it's fine to do it implicitly as driver defaults?

aabadie replied:
> How would removing the header file lead to this application not working with a board having an e-paper display?

No, I meant that keeping the header file you added would forbid this: it will fail with the CI (because of redefined defines).

> I don't think it makes any sense: my personal setup can't be included explicitly, but it's fine to do it implicitly as driver defaults?

I said for convenience with your setup. Since you are contributing this driver, then it's ok to have these defaults. And other users (or boards support) are still allowed to provide their own setup. In the current state of this PR, this is not possible without modifying RIOT (which is annoying).

silkeh replied:
> No, I meant that keeping the header file you added would forbid this: it will fail with the CI (because of redefined defines).

So what would be the proper way to have these overrides?

aabadie replied:
You could use CFLAGS from the command line. But I would avoid any custom setup specific configuration in a RIOT application, this is not generic and cannot be reused in another setup.
If you set the driver default params with your preferred values (let's say it's a privilege given to the initial contributor), then you won't have to change anything. But if someone wants to reuse this application for its needs with a different setup, he can add CFLAGS to the build command line without having to change any line of code. If it's a custom setup hardwired on a board (like the stm32l0 disco), then the defines will have to defined in board.h of this board (board.h must be included in `xx_params.h` header and the application can be built as is.

aabadie commented on tests/driver_epd_bw_spi_disp_dev/Makefile:
```diff
@@ -0,0 +1,9 @@
+USEMODULE += epd_bw_spi
+USEMODULE += disp_dev
+
+INCLUDES += -I$(APPDIR)
```
Same comment as above.

I also think that this application won't be technically required once the support for the stm32l0583-discovery on-board display will be added (I have a branch based on this PR for this): the default `tests/disp_dev` application will cover this use case.

aabadie replied:
> I also think that this application won't be technically required once the support for the stm32l0583-discovery on-board display will be added

Note that I'm not asking to remove it here, just saying that it should be possible to reuse `tests/disp_dev` for the same purpose. And after looking a second time, some changes must be made in `tests/disp_dev` first, so can be done later.

aabadie commented on tests/driver_epd_bw_spi_disp_dev/main.c:
```diff
@@ -0,0 +1,78 @@
+    xtimer_ticks32_t last_wakeup = xtimer_now();
+
+    while (1) {
+        /* Use full refreshes for the full-size logo */
+        epd_bw_spi_set_partial(dev, false);
```
Exactly what I said above: do you expect people to call driver specific functions in an application that is using a generic API ?
But now that I see how you would like to use it, better check for the bounds used with the disp_dev `map` function and call the `epd_bw_spi_update_part`/`epd_bw_spi_update_full` accordingly.

silkeh replied:
Yes, because the `disp_dev` API doesn't cover this fundamental property of e-paper displays.

> better check for the bounds used with the disp_dev map function and call the epd_bw_spi_update_part/epd_bw_spi_update_full accordingly.

The bounds don't necessarily have anything to do with wanting to do a full/partial refresh. The main reason to do a full refresh is to resolve the ghosting, otherwise a partial refresh will always give a better experience (and lower power draw).

aabadie replied:
> The main reason to do a full refresh is to resolve the ghosting, otherwise a partial refresh will always give a better experience (and lower power draw).

Ok, I see. What about we use partial refresh all the time with disp_dev for now (and until we find a better solution with disp_dev) ? Extending the disp_dev interface for this use case is a possibility.

silkeh replied:
What do you think about [the suggestion here](https://github.com/RIOT-OS/RIOT/pull/12509#discussion_r635155942)?
> Maybe instead of a `refresh_partial` flag it should be a `refresh_auto` flag that uses an internal counter to perform a full refresh once every n refreshes?

aabadie reviewed (changes requested):
I made another pass on this PR and I think there are genericity issues with this PR. It would be very nice if we could merge this but I would definitely remove the setup specific from the application directory and better handle the partial/full refresh logic.
I also think that the disp_dev related test application could be removed: RIOT is already supporting a board with an on-board e-ink paper display and it could be used directly with `tests/disp_dev` (with some minor Makefile changes to this application).

See [this branch](https://github.com/aabadie/RIOT/tree/pr/boards/stm32l0528-disco-eink-paper) and especially commits https://github.com/aabadie/RIOT/commit/36e223c4bd5fc54ae61f9789854282ba2a99120d and https://github.com/aabadie/RIOT/commit/9e4f3c081c1fb2ea901954a359d682f3373ab36f.

silkeh commented on drivers/epd_bw_spi/epd_bw_spi_disp_dev.c:
```diff
@@ -0,0 +1,128 @@
+    assert(false);
+}
+
+void epd_bw_spi_set_partial(disp_dev_t *disp_dev, bool partial)
+{
+    ((epd_bw_spi_t *)disp_dev)->refresh_partial = partial;
+}
```
Always using a partial refresh will lead to horrible ghosting, so I don't think it's a good idea to always perform partial refreshes. On the other hand, always doing full refreshes results in a suboptimal display experience and increased power use.

Maybe instead of a `refresh_partial` flag it should be a `refresh_auto` flag that uses an internal counter to perform a full refresh once every `n` refreshes?

aabadie replied:
I don't have a strong opinion on this. If this logic is implemented within the disp_dev maps function, then why not. The value of `n` could be an overridable define.

aabadie commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,329 @@
+int epd_bw_spi_init(epd_bw_spi_t *dev,
+                    const epd_bw_spi_params_t *params,
+                    const epd_bw_spi_controller_t *controller,
+                    uint8_t size_x, uint16_t size_y,
```
size_x and size_y are set via the params (and there is a mismatch between values set in the different controller structs and these values). Same for the controller, there could be one `epd_bw_spi_controller_t` attribute in the params.
Then the signature of the `epd_bw_spi_init` function would be consistent with most of other drivers: `
int epd_bw_spi_init(epd_bw_spi_t *dev, const epd_bw_spi_params_t *params);`.

silkeh replied:
I described my idea for this in [this comment](https://github.com/RIOT-OS/RIOT/pull/12509#issuecomment-826296194).

aabadie replied:
I see, this looks nicer but not from the params struct static initializer point of view. I think a flat struct is ok here:

```c
/**
 * @brief   Driver parameters.
 */
typedef struct {
    spi_t spi;          /**< SPI device that the display is connected to */
    spi_clk_t spi_clk;  /**<; SPI clock speed to use */
    gpio_t cs_pin;      /**<; pin connected to the CHIP SELECT line */
    gpio_t dc_pin;      /**< pin connected to the DC line */
    gpio_t rst_pin;     /**< pin connected to the reset line */
    gpio_t busy_pin;    /**< pin connected to the busy line */
    bool busy_value;    /**< expected value for the busy pin */
    bool dummy;         /**< if device requires a dummy cycle before read */
    epd_bw_spi_entry_mode_t entry_mode;
    uint16_t size_x;    /**< supported number of horizontal pixels */
    uint16_t size_y;    /**< supported number of vertical pixels */
    epd_bw_spi_controller_t controller; /**< E-paper controller configuration */
} epd_bw_spi_params_t;
```

And then in the params:

```c
/**
 * @brief    Default controller definition
 */      
#ifndef EPD_BW_SPI_CONTROLLER
#define EPD_BW_SPI_CONTROLLER EPD_BW_SPI_CONTROLLER_IL3829
#endif

#define EPD_BW_SPI_PARAMS           { .spi = EPD_BW_SPI_PARAM_SPI, \
                                          .spi_clk = EPD_BW_SPI_PARAM_SPI_CLK, \
                                          .cs_pin = EPD_BW_SPI_PARAM_CS, \
                                          .dc_pin = EPD_BW_SPI_PARAM_DC, \
                                          .rst_pin = EPD_BW_SPI_PARAM_RST, \
                                          .busy_pin = EPD_BW_SPI_PARAM_BUSY, \      
                                          .busy_value = EPD_BW_SPI_PARAM_BUSY_VAL, \
                                          .dummy = EPD_BW_SPI_PARAM_DUMMY, \
                                          .entry_mode = EPD_BW_SPI_PARAM_ENTRY_MODE, \
                                          .size_x = EPD_BW_SPI_PARAM_DISPLAY_X, \
                                          .size_y = EPD_BW_SPI_PARAM_DISPLAY_Y, \
                                          .controller = EPD_BW_SPI_PARAM_CONTROLLER, \
}
```

aabadie replied:
Ok, you can keep the X/Y size hard coded for each controller, but X/Y attributes are still needed in the params struct, as showed above. In the code, a sanity check/assert could be added (maybe there's one already, I haven't checked!) to verify that the size_x/size_y configured in the params are within the bounds supported by the controller.

aabadie commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,329 @@
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+void epd_bw_spi_cmd_start(epd_bw_spi_params_t *p, uint8_t cmd, bool cont)
```
It looks weird that you use a pointer to the `epd_bw_spi_params_t` and not to the `epd_bw_spi_t`.

silkeh replied:
So.?

fjmolinas commented:
@aabadie @silkeh can we unblock this one somehow? I'm interested in using this one, so if help is needed let me know, I should get a e-paper on monday and be able to play with this as well.

silkeh commented on drivers/Makefile.dep:
```diff
@@ -247,6 +247,11 @@ ifneq (,$(filter encx24j600,$(USEMODULE)))
   USEMODULE += xtimer
 endif
 
+ifneq (,$(filter epd_bw_spi,$(USEMODULE)))
```
This doesn't work?

aabadie replied:
> This seems to work after a rebase on current master.

Maybe but this is no longer the way to do.

bergzand replied:
> Maybe but this is no longer the way to do.

@aabadie Mind adding a pointer to the current documented way to do things here?

aabadie replied:
It seems that you found your way, great! The current version in this PR is correct. This comment was on an outdated file.

silkeh commented:
> @aabadie @silkeh can we unblock this one somehow? I'm interested in using this one, so if help is needed let me know, I should get a e-paper on monday and be able to play with this as well.

I think I have addressed all concerns, but remind me if I missed anything.

OlegHahm commented:
@aabadie, ping!

OlegHahm commented:
Apart from the failing tests is there anything else missing? @aabadie?

bergzand commented on drivers/include/epd_bw_spi.h:
```diff
@@ -0,0 +1,262 @@
+ * This setting affect the automatic increment/decrement of the address counters.
+ */
+typedef enum {
+    EPD_BW_SPI_Y_DEC_X_DEC  = 0b00, /**< Y decrement, X decrement */
+    EPD_BW_SPI_Y_DEC_X_INC  = 0b01, /**< Y decrement, X increment */
+    EPD_BW_SPI_Y_INC_X_DEC  = 0b10, /**< Y increment, X decrement */
+    EPD_BW_SPI_Y_INC_X_INC  = 0b11, /**< Y increment, X increment */
```
Standard C unfortunately doesn't define binary constants. Could you change these to use hex values instead?

bergzand commented on drivers/include/epd_bw_spi.h:
```diff
@@ -0,0 +1,262 @@
+ * @ingroup     drivers_actuators
+ * @brief       Device driver for black/white e-ink/e-paper SPI displays.
+ *
+ * TODO
```
How about either removing this line or fixing the `TODO` here? :)

bergzand commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,349 @@
+    }
+    else {
+        DEBUG("[epd_bw_spi] wait: for %" PRIu32 " microseconds\n", usec);
+        xtimer_usleep(usec);
```
This needs an include header and a dependency.

If you want to you could migrate this to ztimer. include `ztimer.h` and use a [`ztimer_sleep()`](https://api.riot-os.org/group__sys__ztimer.html#gade98636e198f2d571c8acd861d29d360) here and depend on the relevant `ztimer_{usec,msec,sec}` in the driver.

bergzand commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,349 @@
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+void epd_bw_spi_cmd_start(epd_bw_spi_params_t *p, uint8_t cmd, bool cont)
```
Does this function need to be public (and exported in the `_internal.h` header)? If not please change it to `static`.

bergzand commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,349 @@
+    gpio_set(p->dc_pin);
+}
+
+void epd_bw_spi_write_cmd(epd_bw_spi_params_t *p, uint8_t cmd,
```
Does this function need to be public (and exported in the `_internal.h` header)? If not please change it to `static`.

bergzand commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,349 @@
+    spi_release(p->spi);
+}
+
+void epd_bw_spi_read_cmd(epd_bw_spi_params_t *p, uint8_t cmd, uint8_t *params,
```
Does this function need to be public (and exported in the `_internal.h` header)? If not please change it to `static`.

bergzand commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,349 @@
+    spi_release(p->spi);
+}
+
+void epd_bw_spi_wait(epd_bw_spi_params_t *p, uint32_t usec)
```
Does this function need to be public (and exported in the `_internal.h` header)? If not please change it to `static`.

bergzand reviewed (changes requested):
Few issues from my side.

Other things that are not pressing and definitely something that could be done by a maintainer in a follow up:
 - Migrate the test application to ztimer
 - refactor all `!= GPIO_UNDEF` to use [`gpio_is_valid()`](https://doc.riot-os.org/group__drivers__periph__gpio.html#ga0c3767ed80df4ad880e5bb09520787ed)

silkeh commented:
I have rebased on master, fixed all conflicts (3 years of conflicts :cry:).
There are two remaining issues I need some help with:

- The CI claims that the block of SPI commands in `epd_bw_spi_internal.h` is undocumented, but it is? :confused: 
- The `uint16_t*` cast needed for `disp_dev` leads to a `cast-align` error:
  ```
  tests/driver_epd_bw_spi_disp_dev/main.c:76:48: error: cast increases required alignment of target type [-Werror=cast-align]
     76 |                 disp_dev_map(dev, &small_area, (uint16_t *)riot_logo_32);
  ```

bergzand commented:
> The CI claims that the block of SPI commands in `epd_bw_spi_internal.h` is undocumented, but it is? confused

Technically it is, but I think adding documentation to the individual commands doesn't clarify much. You can add an exclude to `dist/tools/doccheck/generic_exclude_patterns` to silence the doccheck for this:
```
warning: Member EPD_BW_SPI_CMD_[A-Z0-9_]* \(macro definition\) of
```



> The `uint16_t*` cast needed for `disp_dev` leads to a `cast-align` error:

The quick and ugly solution is to cast first to a `void*` to silence the compiler warning. To be honest this is a flaw in the disp_dev api to assume a color format, but I don't see a solution besides changing the disp_dev api to something more flexible.

bergzand commented on tests/driver_epd_bw_spi_disp_dev/main.c:
```diff
@@ -0,0 +1,84 @@
+        disp_dev_map(dev, &large_area, (uint16_t *)riot_icon_200);
+
+        /* Draw small RIOT logos on the display */
+        for (int y = 0; y < disp_dev_width(dev); y += riot_32_height) {
+            for (int x = 0; x < disp_dev_height(dev); x += 32) {
```
```suggestion
        for (unsigned y = 0; y < disp_dev_width(dev); y += riot_32_height) {
            for (unsigned x = 0; x < disp_dev_height(dev); x += 32) {
```

Or maybe even `uint16_t`

bergzand commented on drivers/include/epd_bw_spi.h:
```diff
@@ -0,0 +1,273 @@
+    gpio_t dc_pin;                          /**< pin connected to the DC line */
+    gpio_t rst_pin;                         /**< pin connected to the reset line (optional) */
+    gpio_t busy_pin;                        /**< pin connected to the busy line (optional) */
+    bool busy_value;                        /**< expected value for the busy pin */
```
Another documentation nitpick:

```suggestion
    bool busy_value;                        /**< expected value for the busy pin when the display is busy */
```

bergzand commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,339 @@
+{
+    if (gpio_is_valid(p->busy_pin)) {
+        DEBUG("[epd_bw_spi] wait: for busy bin\n");
+        while (gpio_read(p->busy_pin)) {}
```
```suggestion
        while (gpio_read(p->busy_pin) == p->busy_value) {}
```

bergzand commented on drivers/epd_bw_spi/epd_bw_spi.c:
```diff
@@ -0,0 +1,339 @@
+{
+    uint8_t data[] = { 0xC0 }; /* Enable Clock Signal, Enable CP */
+
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DISPLAY_UPDATE_CONTROL_2,
+                         data, sizeof data);
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_MASTER_ACTIVATION, NULL,
+                         0);
+    epd_bw_spi_wait(&dev->params, EPD_BW_SPI_WAIT_ACTIVATION);
```
And another nitpicky comment.
This pattern of sending a command to `EPD_BW_SPI_CMD_DISPLAY_UPDATE_CONTROL_2`, then writing `EPD_BW_SPI_CMD_DISPLAY_UPDATE_CONTROL_2` with finally the `epd_bw_spi_wait` occurs multiple times. Does it make sense to refactor that in a local function as cleanup?

bergzand commented:
Please squash!

bergzand reviewed (approved):
Ack, thank you for the contribution and your patience with this PR!

bors merge

bergzand commented:
Both test applications need a file called `Makefile.ci` with the following content:
```
BOARD_INSUFFICIENT_MEMORY := \
    arduino-duemilanove \
    arduino-nano \
    arduino-uno \
    atmega328p-xplained-mini \
    atmega328 \
    #
```

Feel free to force push these changes

bergzand commented:
Lets try this again

bors merge

benpicco commented on tests/driver_epd_bw_spi/Makefile.ci:
```diff
@@ -0,0 +1,7 @@
+BOARD_INSUFFICIENT_MEMORY := \
```
You can generate this with

    make create-Makefile.ci

bergzand commented on tests/driver_epd_bw_spi/Makefile.ci:
```diff
@@ -0,0 +1,7 @@
+BOARD_INSUFFICIENT_MEMORY := \
+    arduino-duemilanove \
+    arduino-nano \
+    arduino-uno \
+    atmega328p-xplained-mini \
+    atmega328 \
+    #
```
As generated by `make create-Makefile.ci` (and using the docker container)
```suggestion
BOARD_INSUFFICIENT_MEMORY := \
    arduino-duemilanove \
    arduino-leonardo \
    arduino-nano \
    arduino-uno \
    atmega328p \
    atmega328p-xplained-mini \
    #
```

bergzand commented on tests/driver_epd_bw_spi_disp_dev/Makefile.ci:
```diff
@@ -0,0 +1,7 @@
+BOARD_INSUFFICIENT_MEMORY := \
+    arduino-duemilanove \
+    arduino-nano \
+    arduino-uno \
+    atmega328p-xplained-mini \
+    atmega328 \
+    #
```
As generated by `make create-Makefile.ci` (and using the docker container)
```suggestion
BOARD_INSUFFICIENT_MEMORY := \
    arduino-duemilanove \
    arduino-leonardo \
    arduino-nano \
    arduino-uno \
    atmega328p \
    atmega328p-xplained-mini \
    nucleo-l011k4 \
    samd10-xmini \
    stm32f030f4-demo \
    #
```

bergzand reviewed (approved):
Thanks,

Bors merge

bergzand commented:
Bors merge

bergzand commented:
Thanks for the effort you put into this @silkeh
